File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/util/UIUtils.java
package com.byagowi.persiancalendar.util;

import android.Manifest;
import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Build;
import android.preference.PreferenceManager;
import android.text.Html;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.byagowi.persiancalendar.Constants;
import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.entity.DeviceCalendarEvent;
import com.byagowi.persiancalendar.praytimes.Clock;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;
import androidx.annotation.StyleRes;
import androidx.appcompat.app.AlertDialog;

import static com.byagowi.persiancalendar.Constants.BLUE_THEME;
import static com.byagowi.persiancalendar.Constants.DARK_THEME;
import static com.byagowi.persiancalendar.Constants.LIGHT_THEME;
import static com.byagowi.persiancalendar.Constants.MODERN_THEME;
import static com.byagowi.persiancalendar.Constants.PREF_SHOW_DEVICE_CALENDAR_EVENTS;

public class UIUtils {
    private static final long twoSeconds = TimeUnit.SECONDS.toMillis(2);
    private static long latestToastShowTime = -1;
    private static AudioManager audioManager = null;

    public static void askForCalendarPermission(Activity activity) {
        if (activity == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return;

        new AlertDialog.Builder(activity)
                .setTitle(R.string.calendar_access)
                .setMessage(R.string.phone_calendar_required)
                .setPositiveButton(R.string.continue_button, (dialog, id) -> activity.requestPermissions(new String[]{
                                Manifest.permission.READ_CALENDAR
                        },
                        Constants.CALENDAR_READ_PERMISSION_REQUEST_CODE))
                .setNegativeButton(R.string.cancel, (dialog, id) -> dialog.cancel()).show();
    }

    public static void askForLocationPermission(Activity activity) {
        if (activity == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return;

        new AlertDialog.Builder(activity)
                .setTitle(R.string.location_access)
                .setMessage(R.string.phone_location_required)
                .setPositiveButton(R.string.continue_button, (dialog, id) -> activity.requestPermissions(new String[]{
                                Manifest.permission.ACCESS_COARSE_LOCATION,
                                Manifest.permission.ACCESS_FINE_LOCATION
                        },
                        Constants.LOCATION_PERMISSION_REQUEST_CODE))
                .setNegativeButton(R.string.cancel, (dialog, id) -> dialog.cancel()).show();
    }

    public static void toggleShowDeviceCalendarOnPreference(Context context, boolean enable) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = prefs.edit();
        edit.putBoolean(PREF_SHOW_DEVICE_CALENDAR_EVENTS, enable);
        edit.apply();
    }

    static public String formatDeviceCalendarEventTitle(DeviceCalendarEvent event) {
        String desc = event.getDescription();
        String title = event.getTitle();
        if (!TextUtils.isEmpty(desc))
            title += " (" + Html.fromHtml(event.getDescription()).toString().trim() + ")";

        return title.replaceAll("\\n", " ").trim();
    }

    public static String baseFormatClock(int hour, int minute) {
        return Utils.formatNumber(String.format(Locale.ENGLISH, "%d:%02d", hour, minute));
    }

    public static boolean isRTL(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return context.getResources().getConfiguration().getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
        }
        return false;
    }

    static public String getFormattedClock(Clock clock, boolean forceIn12) {
        boolean in12 = Utils.isClockIn12() || forceIn12;
        if (!in12) return baseFormatClock(clock.getHour(), clock.getMinute());

        int hour = clock.getHour();
        String suffix;
        if (hour >= 12) {
            suffix = Utils.getAmString();
            hour -= 12;
        } else {
            suffix = Utils.getPmString();
        }
        return baseFormatClock(hour, clock.getMinute()) + " " + suffix;
    }

    static public @StringRes
    int getPrayTimeText(String athanKey) {
        switch (athanKey) {
            case "FAJR":
                return R.string.fajr;

            case "DHUHR":
                return R.string.dhuhr;

            case "ASR":
                return R.string.asr;

            case "MAGHRIB":
                return R.string.maghrib;

            case "ISHA":
            default:
                return R.string.isha;
        }
    }

    static public @DrawableRes
    int getPrayTimeImage(String athanKey) {
        switch (athanKey) {
            case "FAJR":
                return R.drawable.fajr;

            case "DHUHR":
                return R.drawable.dhuhr;

            case "ASR":
                return R.drawable.asr;

            case "MAGHRIB":
                return R.drawable.maghrib;

            case "ISHA":
            default:
                return R.drawable.isha;
        }
    }

    @StyleRes
    public static int getThemeFromName(String name) {
        switch (name) {
            case DARK_THEME:
                return R.style.DarkTheme;

            case MODERN_THEME:
                return R.style.ModernTheme;

            case BLUE_THEME:
                return R.style.BlueTheme;

            default:
            case LIGHT_THEME:
                return R.style.LightTheme;
        }
    }

    // https://stackoverflow.com/a/27788209
    static public Uri getDefaultAthanUri(Context context) {
        return Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" +
                context.getResources().getResourcePackageName(R.raw.abdulbasit) + '/' +
                context.getResources().getResourceTypeName(R.raw.abdulbasit) + '/' +
                context.getResources().getResourceEntryName(R.raw.abdulbasit));
    }

    static String getOnlyLanguage(String string) {
        return string.replaceAll("-(IR|AF|US)", "");
    }

    public static void a11yShowToastWithClick(Context context, @StringRes int resId) {
        if (!Utils.isTalkBackEnabled()) return;

        if (audioManager == null) {
            audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        }

        long now = System.currentTimeMillis();
        if (now - latestToastShowTime > twoSeconds) {
            Toast.makeText(context, resId, Toast.LENGTH_SHORT).show();
            // https://stackoverflow.com/a/29423018
            audioManager.playSoundEffect(AudioManager.FX_KEY_CLICK);
            latestToastShowTime = now;
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/util/UIUtils.kt
package com.byagowi.persiancalendar.util

import android.Manifest
import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.os.Build
import android.preference.PreferenceManager
import android.text.Html
import android.text.TextUtils
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import calendar.DateConverter
import com.byagowi.persiancalendar.Constants
import com.byagowi.persiancalendar.Constants.AM_IN_CKB
import com.byagowi.persiancalendar.Constants.AM_IN_PERSIAN
import com.byagowi.persiancalendar.Constants.DARK_THEME
import com.byagowi.persiancalendar.Constants.LIGHT_THEME
import com.byagowi.persiancalendar.Constants.PM_IN_CKB
import com.byagowi.persiancalendar.Constants.PM_IN_PERSIAN
import com.byagowi.persiancalendar.Constants.PREF_THEME
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.databinding.CalendarsCardBinding
import com.byagowi.persiancalendar.databinding.SelectdayFragmentBinding
import com.byagowi.persiancalendar.entity.DeviceCalendarEvent
import com.github.praytimes.Clock
import java.util.*

object UIUtils {
  fun setActivityTitleAndSubtitle(activity: Activity, title: String, subtitle: String) {

    val supportActionBar = (activity as AppCompatActivity).supportActionBar
    if (supportActionBar != null) {
      supportActionBar.title = title
      supportActionBar.subtitle = subtitle
    }
  }

  fun fillCalendarsCard(context: Context, jdn: Long,
                        binding: CalendarsCardBinding, isToday: Boolean) {
    val persianDate = DateConverter.jdnToPersian(jdn)
    val civilDate = DateConverter.jdnToCivil(jdn)
    val hijriDate = DateConverter.civilToIslamic(civilDate, Utils.getIslamicOffset())

    binding.weekDayName.text = Utils.getWeekDayName(civilDate)
    binding.shamsiDateLinear.text = CalendarUtils.toLinearDate(persianDate)
    binding.shamsiDateDay.text = Utils.formatNumber(persianDate.dayOfMonth)
    binding.shamsiDate.text = CalendarUtils.getMonthName(persianDate) + "\n" + Utils.formatNumber(persianDate.year)

    binding.gregorianDateLinear.text = CalendarUtils.toLinearDate(civilDate)
    binding.gregorianDateDay.text = Utils.formatNumber(civilDate.dayOfMonth)
    binding.gregorianDate.text = CalendarUtils.getMonthName(civilDate) + "\n" + Utils.formatNumber(civilDate.year)

    binding.islamicDateLinear.text = CalendarUtils.toLinearDate(hijriDate)
    binding.islamicDateDay.text = Utils.formatNumber(hijriDate.dayOfMonth)
    binding.islamicDate.text = CalendarUtils.getMonthName(hijriDate) + "\n" + Utils.formatNumber(hijriDate.year)

    if (isToday) {
      binding.today.visibility = View.GONE
      binding.todayIcon.visibility = View.GONE
      if (Utils.isIranTime) {
        binding.weekDayName.text = binding.weekDayName.text.toString() + " (" + context.getString(R.string.iran_time) + ")"
      }
    } else {
      binding.today.visibility = View.VISIBLE
      binding.todayIcon.visibility = View.VISIBLE
    }
  }

  fun fillSelectdaySpinners(context: Context, binding: SelectdayFragmentBinding): Int {
    val date = CalendarUtils.getTodayOfCalendar(CalendarUtils.calendarTypeFromPosition(
        binding.calendarTypeSpinner.selectedItemPosition))

    // years spinner init.
    val years = arrayOfNulls<String>(200)
    val startingYearOnYearSpinner = date.year - years.size / 2
    for (i in years.indices) {
      years[i] = Utils.formatNumber(i + startingYearOnYearSpinner)
    }
    binding.yearSpinner.adapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_dropdown_item, years)
    binding.yearSpinner.setSelection(years.size / 2)
    //

    // month spinner init.
    val months = Utils.monthsNamesOfCalendar(date).clone()
    for (i in months.indices) {
      months[i] = months[i] + " / " + Utils.formatNumber(i + 1)
    }
    binding.monthSpinner.adapter = ArrayAdapter(context,
        android.R.layout.simple_spinner_dropdown_item, months)
    binding.monthSpinner.setSelection(date.month - 1)
    //

    // days spinner init.
    val days = arrayOfNulls<String>(31)
    for (i in days.indices) {
      days[i] = Utils.formatNumber(i + 1)
    }
    binding.daySpinner.adapter = ArrayAdapter<String>(context,
        android.R.layout.simple_spinner_dropdown_item, days)
    binding.daySpinner.setSelection(date.dayOfMonth - 1)
    //

    return startingYearOnYearSpinner
  }

  fun askForCalendarPermission(activity: AppCompatActivity) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      activity.requestPermissions(arrayOf(Manifest.permission.READ_CALENDAR),
          Constants.CALENDAR_READ_PERMISSION_REQUEST_CODE)
    }
  }

  fun formatDeviceCalendarEventTitle(event: DeviceCalendarEvent): String {
    val desc = event.description
    var title = event.title
    if (!TextUtils.isEmpty(desc))
      title += " (" + Html.fromHtml(event.description).toString().trim { it <= ' ' } + ")"

    return title.replace("\\n".toRegex(), " ").trim { it <= ' ' }
  }

  internal fun clockToString(hour: Int, minute: Int): String = Utils.formatNumber(String.format(Locale.ENGLISH, "%d:%02d", hour, minute))

  fun isRTL(context: Context): Boolean =
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
        context.resources.configuration.layoutDirection == View.LAYOUT_DIRECTION_RTL
      } else false

  fun getFormattedClock(clock: Clock): String {
    var timeText: String? = null

    var hour = clock.hour
    if (!Utils.isClockIn24) {
      if (hour >= 12) {
        timeText = if (Utils.appLanguage == "ckb")
          PM_IN_CKB
        else
          PM_IN_PERSIAN
        hour -= 12
      } else {
        timeText = if (Utils.appLanguage == "ckb")
          AM_IN_CKB
        else
          AM_IN_PERSIAN
      }
    }

    var result = clockToString(hour, clock.minute)
    if (!Utils.isClockIn24) {
      result = "$result $timeText"
    }
    return result
  }

  @StringRes
  fun getPrayTimeText(athanKey: String): Int = when (athanKey) {
    "FAJR" -> R.string.azan1

    "DHUHR" -> R.string.azan2

    "ASR" -> R.string.azan3

    "MAGHRIB" -> R.string.azan4

    "ISHA" -> R.string.azan5
    else -> R.string.azan5
  }

  @DrawableRes
  fun getPrayTimeImage(athanKey: String): Int = when (athanKey) {
    "FAJR" -> R.drawable.fajr

    "DHUHR" -> R.drawable.dhuhr

    "ASR" -> R.drawable.asr

    "MAGHRIB" -> R.drawable.maghrib

    "ISHA" -> R.drawable.isha

    else -> R.drawable.isha
  }

  fun copyToClipboard(context: Context, text: CharSequence) {
    val clipboardService =
        context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?

    if (clipboardService != null) {
      clipboardService.primaryClip = ClipData.newPlainText("converted date", text)
      Toast.makeText(context, "«" + text + "»\n" + context.getString(R.string.date_copied_clipboard), Toast.LENGTH_SHORT).show()
    }
  }

  fun setTheme(activity: AppCompatActivity) {
    when (PreferenceManager.getDefaultSharedPreferences(activity).getString(PREF_THEME, LIGHT_THEME)) {
      DARK_THEME -> {
        activity.setTheme(R.style.DarkTheme)
        return
      }
      LIGHT_THEME -> activity.setTheme(R.style.LightTheme)
      //            case CLASSIC_THEME:
      //                setTheme(R.style.ClassicTheme);
      //                return;
      else -> activity.setTheme(R.style.LightTheme)
    }
  }

  // https://stackoverflow.com/a/27788209
  private fun resourceToUri(context: Context, resID: Int): Uri =
      Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" +
          context.resources.getResourcePackageName(resID) + '/'.toString() +
          context.resources.getResourceTypeName(resID) + '/'.toString() +
          context.resources.getResourceEntryName(resID))

  fun getDefaultAthanUri(context: Context): Uri = resourceToUri(context, R.raw.abdulbasit)

  internal fun getOnlyLanguage(string: String): String = string.replace("-(IR|AF|US)".toRegex(), "")
}
